### PR TITLE
fix: delete created Kafka if the corresponding command fails

### PR DESCRIPTION
**Pull request by rkhachatryan**
### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")
- [ ] Do these changes have compatibility implications for rollback? If so, ensure that the ksql [command version](https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/Command.java#L41) is bumped.

### CODE REVIEW

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java [hunk 1]
```diff
@@ -172,10 +174,27 @@ public StatementExecutorResponse execute(
...
+            injectorFactory.apply(executionContext, securityContext.getServiceContext()));
+    final ConfiguredStatementWithSideEffects<?> injectedWithSideEffects =
+            injector.injectWithSideEffects(statement);
+    try {
+      return executeInjected(
+            injectedWithSideEffects.getStatement(), statement,
+            executionContext,
+            securityContext);
+    } catch (Exception e) {
+      injector.revertSideEffects(injectedWithSideEffects);
```
Comment (rkhachatryan) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:187]: I'm wondering whether a race condition is possible:
1. Instance A checks the topic, it doesn’t exist, so it creates it
2. Instance B checks the topic, it exists, so it skips creation
3. Instance B fences off instance A
4. Instance A deletes the topic as a result of failure handling
5. Instance B continues execution assuming that the topic exists

If it is possible, we should probably ignore `ProducerFencedException` (assuming that commands always result in this topic being created). This also leaves a very small possibility of not deleting this topic - if the newer producer fails too.

File: ksqldb-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java [hunk 2]
```diff
@@ -50,7 +53,7 @@
  *
  * @see TopicProperties.Builder
  */
-public class TopicCreateInjector implements Injector {
+public class TopicCreateInjector implements InjectorWithSideEffects {
```
Comment (rkhachatryan) [inline ksqldb-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java:56]: I assume that `DistributingExecutor` is the only code path that is susceptible to the problem.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java [hunk 3]
```diff
@@ -172,10 +174,27 @@ public StatementExecutorResponse execute(
...
-        .inject(statement);
+    final InjectorWithSideEffects injector = InjectorWithSideEffects.wrap(
+            injectorFactory.apply(executionContext, securityContext.getServiceContext()));
+    final ConfiguredStatementWithSideEffects<?> injectedWithSideEffects =
+            injector.injectWithSideEffects(statement);
+    try {
+      return executeInjected(
+            injectedWithSideEffects.getStatement(), statement,
+            executionContext,
+            securityContext);
```
Comment (rkhachatryan) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:185]: I assume that **any** error inside `executeInjected` should result in topic deletion if it was created (apart from `ProducerFencedException` probably - discussed separately).

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java (hunk 1 above)
Comment (mjsax) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:187]: As discussed offline: there is a ton of race condition and focusing on the "happy cleanup path" is a good step forward. -- Making it absolutely bullet proof seems to be "impossible" anyway (ie, we would need to change Kafka to get an atomic "create topics and write into command-topic" API)

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java [hunk 4]
```diff
@@ -172,10 +174,27 @@ public StatementExecutorResponse execute(
...
-    final ConfiguredStatement<?> injected = injectorFactory
-        .apply(executionContext, securityContext.getServiceContext())
-        .inject(statement);
+    final InjectorWithSideEffects injector = InjectorWithSideEffects.wrap(
+            injectorFactory.apply(executionContext, securityContext.getServiceContext()));
+    final ConfiguredStatementWithSideEffects<?> injectedWithSideEffects =
+            injector.injectWithSideEffects(statement);
+    try {
+      return executeInjected(
+            injectedWithSideEffects.getStatement(), statement,
```
Comment (mjsax) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:183]: nit: move `statement` to next line

**Review by mjsax — APPROVED** [review]

**Pull request merged**